File: packages/go-agentenkit/state_transition_test.go
package agentenkit_test

import (
	"context"
	"errors"
	"testing"
	"time"

	agentenkit "github.com/eadwinCode/agentic-kit/packages/go-agentenkit"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/memory"
	pgstorage "github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/postgres"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/core"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/ports"
)

// Workstream C: every run state change is one compare-and-set on the state
// AND the run that owns the thread. The same cases run in the TS package
// (test/state-transition.test.ts).

// transitionContract is what every Storage adapter's Transition must do.
func transitionContract(t *testing.T, s ports.Storage) {
	t.Helper()
	ctx := context.Background()
	sc := ports.StorageContext{}
	th, err := s.Threads().Create(ctx, ports.ThreadInit{}, sc)
	if err != nil {
		t.Fatal(err)
	}
	step := func(name string, tr ports.ThreadTransition, want bool, state ports.ExecutionState) {
		t.Helper()
		got, err := s.Threads().Transition(ctx, th.ID, tr, sc)
		if err != nil {
			t.Fatalf("%s: %v", name, err)
		}
		if got != want {
			t.Fatalf("%s: won=%v, want %v", name, got, want)
		}
		now, _ := s.Threads().Get(ctx, th.ID, sc)
		if now.State != state {
			t.Fatalf("%s: state %s, want %s", name, now.State, state)
		}
	}
	Q, R, W, C := ports.StateQueued, ports.StateRunning, ports.StateWaitingForInput, ports.StateCancelled
	step("admission records the run", ports.ThreadTransition{From: []ports.ExecutionState{ports.StateIdle}, To: Q, NewRunID: "r1"}, true, Q)
	step("wrong state loses", ports.ThreadTransition{From: []ports.ExecutionState{R}, To: ports.StateCompleted, RunID: "r1"}, false, Q)
	step("another run loses", ports.ThreadTransition{From: []ports.ExecutionState{Q}, To: R, RunID: "r2"}, false, Q)
	step("its own run wins", ports.ThreadTransition{From: []ports.ExecutionState{Q}, To: R, RunID: "r1"}, true, R)
	step("any of several states", ports.ThreadTransition{From: []ports.ExecutionState{Q, R, W}, To: C, RunID: "r1"}, true, C)
	step("a new run is admitted", ports.ThreadTransition{From: []ports.ExecutionState{C}, To: Q, NewRunID: "r2"}, true, Q)
	step("the replaced run can no longer move it", ports.ThreadTransition{From: []ports.ExecutionState{Q}, To: C, RunID: "r1"}, false, Q)
	step("no run named matches any run", ports.ThreadTransition{From: []ports.ExecutionState{Q}, To: R}, true, R)

	// A thread from before the run was recorded matches any run.
	legacy, _ := s.Threads().Create(ctx, ports.ThreadInit{}, sc)
	if won, err := s.Threads().Transition(ctx, legacy.ID, ports.ThreadTransition{From: []ports.ExecutionState{ports.StateIdle}, To: ports.StateFailed, RunID: "x"}, sc); err != nil || !won {
		t.Fatalf("a thread with no run recorded matches any run: %v %v", won, err)
	}
	if won, _ := s.Threads().Transition(ctx, "no-such-thread", ports.ThreadTransition{From: []ports.ExecutionState{ports.StateIdle}, To: Q}, sc); won {
		t.Fatal("an unknown thread never matches")
	}
}

func TestTransition_MemoryStorage(t *testing.T) { transitionContract(t, memory.NewStorage()) }
func TestTransition_SqliteStorage(t *testing.T) { transitionContract(t, openSqlite(t)) }
func TestTransition_PostgresStorage(t *testing.T) {
	db := openPostgres(t)
	ctx := context.Background()
	for _, tbl := range []string{"usage", "events", "messages", "threads"} {
		_, _ = db.ExecContext(ctx, "DROP TABLE IF EXISTS tr_"+tbl)
	}
	s, err := pgstorage.New(ctx, db, pgstorage.WithPrefix("tr_"))
	if err != nil {
		t.Fatal(err)
	}
	transitionContract(t, s)
}

func TestState_AFinishLosesToAStopThatLandedFirst(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "hi"})
	deps := h.rt.Ports(nil)
	// A worker picked the run up...
	if won, err := core.Transition(h.ctx, deps, ran.ThreadID, core.StateChange{
		From: []ports.ExecutionState{ports.StateQueued}, To: ports.StateRunning, RunID: ran.RunID,
	}); err != nil || !won {
		t.Fatalf("pickup: %v %v", won, err)
	}
	// ...the user stopped it...
	if res, err := chat.Stop(h.ctx, ran.ThreadID, nil); err != nil || !res.Accepted {
		t.Fatalf("stop: %+v %v", res, err)
	}
	// ...and then the worker finished, having read RUNNING before the stop.
	if err := agentenkit.Finalize(h.ctx, deps, nil, ran.ThreadID, agentenkit.FinalizeInput{
		State: ports.StateCompleted, StopReason: "completed", RunID: ran.RunID,
	}); err != nil {
		t.Fatal(err)
	}
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateCancelled, "the stop stands")
	mustEqual(t, h.kvGet(agentenkit.StateKey(ran.ThreadID)), "CANCELLED", "on the hot cache too")
	mustStrings(t, h.states(ran.ThreadID), []string{"QUEUED", "CANCELLED"}, "no COMPLETED after the stop")
	rec, _ := h.admin.Runs().Get(h.ctx, ran.RunID)
	mustEqual(t, rec.State, agentenkit.StateCancelled, "the record keeps the stop")
}

func TestStop_AQueuedRunIsCancelledAndItsJobDoesNothing(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "hi"})
	res, err := chat.Stop(h.ctx, ran.ThreadID, nil)
	if err != nil || !res.Accepted {
		t.Fatalf("a queued run can be stopped: %+v %v", res, err)
	}
	h.handleNext(t)
	mustEqual(t, h.model.Calls(), 0, "the job does nothing")
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateCancelled, "CANCELLED")
}

func TestPark_AStoppedRunParksNothing(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "hi"})
	deps := h.rt.Ports(nil)
	_, _ = core.Transition(h.ctx, deps, ran.ThreadID, core.StateChange{
		From: []ports.ExecutionState{ports.StateQueued}, To: ports.StateRunning, RunID: ran.RunID,
	})
	if _, err := chat.Stop(h.ctx, ran.ThreadID, nil); err != nil {
		t.Fatal(err)
	}
	// A tool that was mid-call when the stop landed now tries to park.
	if err := core.ParkForApproval(h.ctx, deps, core.ParkInput{
		ThreadID: ran.ThreadID, ToolCallID: "c1", ToolName: "send",
		Resume: ports.ResumeInfo{Agent: "chat", Model: "gpt-4o", RunID: ran.RunID},
	}); err != nil {
		t.Fatal(err)
	}
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateCancelled, "still CANCELLED")
	mustEqual(t, len(h.events(ran.ThreadID, "INPUT_REQUIRED")), 0, "no approval request")
	mustEqual(t, h.queue.Len(), 1, "no expiry job (only the original dispatch)")
}

func TestRetry_AReplacedRunsFailureSpendsNoAttempt(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}))
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "hi"})
	job, _ := h.queue.Shift()
	_, _ = h.kv.Set(h.ctx, agentenkit.RunIDKey(ran.ThreadID), "a-newer-run", ports.SetOptions{})
	err := chat.ExecuteWithPolicy(h.ctx, agentenkit.ExecuteInput{ThreadID: job.ThreadID, RunID: job.RunID, Model: job.Model},
		&agentenkit.Policy{Exec: func(context.Context, ports.RuntimePorts, *agentenkit.RegisteredAgent, agentenkit.ExecuteInput) (agentenkit.ExecuteOutcome, error) {
			return "", errors.New("boom")
		}})
	if err != nil {
		t.Fatal(err)
	}
	mustEqual(t, h.kvGet(agentenkit.AttemptsKey(ran.RunID)), "", "no attempt spent")
	mustEqual(t, h.queue.Len(), 0, "nothing retried")
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateQueued, "the thread is left to the newer run")
}

func TestRetry_AFailedRunWaitsQueued(t *testing.T) {
	h := makeRuntime(t, scripted(step{err: errBoom}, step{text: "ok"}), func(c *agentenkit.AgentConfig) {
		c.RunRetryBackoff = time.Millisecond
	})
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "hi"})
	h.handleNext(t)
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateQueued, "QUEUED while it waits to retry")
	mustStrings(t, h.states(ran.ThreadID), []string{"QUEUED", "RUNNING", "QUEUED"}, "and says so")
	mustEqual(t, h.kvGet(agentenkit.AttemptsKey(ran.RunID)), "1", "one attempt, counted per run")
	h.handleNext(t)
	mustEqual(t, h.thread(t, ran.ThreadID).State, agentenkit.StateCompleted, "the retry completes it")
}

File: packages/agentenkit/test/state-transition.test.ts
import { describe, expect, it } from 'bun:test';
import { Database } from 'bun:sqlite';
import { simulateReadableStream } from 'ai';
import type { LanguageModelV1StreamPart } from '@ai-sdk/provider';
import { MockLanguageModelV1 } from 'ai/test';
import { setupAgentCore } from '../src/runtime.js';
import { MemoryAdminStore } from '../src/admin/memory.js';
import { MemoryBus, MemoryKv, MemoryQueue, MemoryStorage } from '../src/adapters/memory.js';
import { SqliteStorage, type SqliteLike } from '../src/adapters/sqlite.js';
import { bindStorage } from '../src/core/state.js';
import { finalize } from '../src/core/engine.js';
import { parkForApproval } from '../src/core/hitl.js';
import { transition } from '../src/core/publish.js';
import { attemptsKey, runIdKey } from '../src/core/keys.js';
import { resolveConfig, type AgentConfig, type ExecutionState } from '../src/core/types.js';
import type { Storage } from '../src/ports/storage.js';
import type { RuntimePorts } from '../src/ports/runtime.js';

// Workstream C: every run state change is one compare-and-set on the state
// AND the run that owns the thread. The same cases run in the Go package
// (state_transition_test.go).

/** What every Storage adapter's `transition` must do. */
async function transitionContract(s: Storage) {
  const ctx = { state: {} };
  const th = await s.threads.create({}, ctx);
  const step = async (
    name: string,
    t: { from: ExecutionState[]; to: ExecutionState; runId?: string; newRunId?: string },
    won: boolean,
    state: ExecutionState,
  ) => {
    expect([name, await s.threads.transition(th.id, t, ctx)]).toEqual([name, won]);
    expect([name, (await s.threads.get(th.id, ctx))!.state]).toEqual([name, state]);
  };
  await step('admission records the run', { from: ['IDLE'], to: 'QUEUED', newRunId: 'r1' }, true, 'QUEUED');
  await step('wrong state loses', { from: ['RUNNING'], to: 'COMPLETED', runId: 'r1' }, false, 'QUEUED');
  await step('another run loses', { from: ['QUEUED'], to: 'RUNNING', runId: 'r2' }, false, 'QUEUED');
  await step('its own run wins', { from: ['QUEUED'], to: 'RUNNING', runId: 'r1' }, true, 'RUNNING');
  await step('any of several states', { from: ['QUEUED', 'RUNNING', 'WAITING_FOR_INPUT'], to: 'CANCELLED', runId: 'r1' }, true, 'CANCELLED');
  await step('a new run is admitted', { from: ['CANCELLED'], to: 'QUEUED', newRunId: 'r2' }, true, 'QUEUED');
  await step('the replaced run can no longer move it', { from: ['QUEUED'], to: 'CANCELLED', runId: 'r1' }, false, 'QUEUED');
  await step('no run named matches any run', { from: ['QUEUED'], to: 'RUNNING' }, true, 'RUNNING');

  // A thread from before the run was recorded matches any run.
  const legacy = await s.threads.create({}, ctx);
  expect(await s.threads.transition(legacy.id, { from: ['IDLE'], to: 'FAILED', runId: 'x' }, ctx)).toBe(true);
  expect(await s.threads.transition('no-such-thread', { from: ['IDLE'], to: 'QUEUED' }, ctx)).toBe(false);
}

describe('Storage.threads.transition', () => {
  it('MemoryStorage', () => transitionContract(new MemoryStorage()));
  it('SqliteStorage', () => transitionContract(new SqliteStorage(new Database(':memory:') as unknown as SqliteLike)));
});

function model(steps: Array<'fail' | 'ok'>, before?: () => Promise<void>) {
  let call = 0;
  return new MockLanguageModelV1({
    provider: 'mock',
    modelId: 'mock',
    doStream: async () => {
      const s = steps[Math.min(call++, steps.length - 1)];
      await before?.();
      if (s === 'fail') throw new Error('boom');
      const chunks: LanguageModelV1StreamPart[] = [
        { type: 'text-delta', textDelta: 'ok' },
        { type: 'finish', finishReason: 'stop', usage: { promptTokens: 10, completionTokens: 5 } },
      ];
      return { stream: simulateReadableStream({ chunks }), rawCall: { rawPrompt: null, rawSettings: {} } };
    },
  });
}

async function makeRuntime(m: any, config: Partial<AgentConfig> = {}) {
  const storage = new MemoryStorage();
  const bus = new MemoryBus();
  const queue = new MemoryQueue();
  const kv = new MemoryKv();
  const admin = new MemoryAdminStore();
  const resolved = resolveConfig({ stopPollMs: 20, ...config });
  const runtime = await setupAgentCore({
    storage, bus, queue, kv, admin,
    resolveModel: () => ({ instance: () => m, contextWindow: 128_000 }),
    config: resolved,
  });
  const ports: RuntimePorts = {
    storage: bindStorage(storage, { state: {} }), bus, queue, kv, admin, config: resolved,
    resolveModel: () => ({ instance: () => m, contextWindow: 128_000 }),
  };
  const chat = runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o' });
  const states = (threadId: string) =>
    bus.published
      .filter((e) => e.threadId === threadId && e.type === 'STATE_CHANGE')
      .map((e) => (e.payload as any).state);
  return { runtime, chat, storage, bus, queue, kv, admin, ports, states };
}

describe('state changes are compare-and-set', () => {
  it('a finish loses to a stop that landed first', async () => {
    const r = await makeRuntime(model(['ok']));
    const ran = await r.chat.run({ prompt: 'hi' });
    // A worker picked the run up…
    expect(await transition(r.ports, ran.threadId, { from: ['QUEUED'], to: 'RUNNING', runId: ran.runId })).toBe(true);
    // …the user stopped it…
    expect((await r.chat.stop(ran.threadId)).accepted).toBe(true);
    // …and then the worker finished, having read RUNNING before the stop.
    await finalize(r.ports, undefined as never, ran.threadId, {
      state: 'COMPLETED', stopReason: 'completed', tokensUsed: 0,
      attribution: { inputTokens: 0, cachedInputTokens: 0, outputTokens: 0, totalTokens: 0 },
      runId: ran.runId,
    });
    expect((await r.storage.threads.get(ran.threadId))!.state).toBe('CANCELLED');
    expect(await r.kv.get(`agent:state:${ran.threadId}`)).toBe('CANCELLED');
    expect(r.states(ran.threadId)).toEqual(['QUEUED', 'CANCELLED']); // no COMPLETED after the stop
    expect((await r.admin.runs.get(ran.runId!))!.state).toBe('CANCELLED');
  });

  it('a queued run is cancelled and its job does nothing', async () => {
    let calls = 0;
    const r = await makeRuntime(model(['ok'], async () => void calls++));
    const ran = await r.chat.run({ prompt: 'hi' });
    expect((await r.chat.stop(ran.threadId)).accepted).toBe(true);
    await r.runtime.worker.handleJob(r.queue.items[0]!);
    expect(calls).toBe(0);
    expect((await r.storage.threads.get(ran.threadId))!.state).toBe('CANCELLED');
  });

  it('a stopped run parks nothing', async () => {
    const r = await makeRuntime(model(['ok']));
    const ran = await r.chat.run({ prompt: 'hi' });
    await transition(r.ports, ran.threadId, { from: ['QUEUED'], to: 'RUNNING', runId: ran.runId });
    await r.chat.stop(ran.threadId);
    // A tool that was mid-call when the stop landed now tries to park.
    await parkForApproval(r.ports, {
      threadId: ran.threadId, toolCallId: 'c1', toolName: 'send', args: {},
      resume: { agent: 'chat', model: 'gpt-4o', runId: ran.runId },
    } as any);
    expect((await r.storage.threads.get(ran.threadId))!.state).toBe('CANCELLED');
    expect(r.bus.published.some((e) => e.type === 'INPUT_REQUIRED')).toBe(false);
    expect(r.queue.items).toHaveLength(1); // no expiry job (only the original dispatch)
  });

  it("a replaced run's failure spends no attempt", async () => {
    let threadId = '';
    let kv!: MemoryKv;
    // The run is replaced while its model call is in flight, then the call fails.
    const r = await makeRuntime(model(['fail'], async () => {
      await kv.set(runIdKey(threadId), 'a-newer-run');
    }));
    kv = r.kv;
    const ran = await r.chat.run({ prompt: 'hi' });
    threadId = ran.threadId;
    await r.runtime.worker.handleJob(r.queue.items[0]!);
    expect(await r.kv.get(attemptsKey(ran.runId!))).toBeNull(); // no attempt spent
    expect(r.queue.items).toHaveLength(1); // nothing retried
  });

  it('a failed run waits QUEUED for its retry', async () => {
    const r = await makeRuntime(model(['fail', 'ok']), { runRetryBackoffMs: 1 });
    const ran = await r.chat.run({ prompt: 'hi' });
    await r.runtime.worker.handleJob(r.queue.items[0]!);
    expect((await r.storage.threads.get(ran.threadId))!.state).toBe('QUEUED');
    expect(r.states(ran.threadId)).toEqual(['QUEUED', 'RUNNING', 'QUEUED']);
    expect(await r.kv.get(attemptsKey(ran.runId!))).toBe('1'); // counted per run
    await r.runtime.worker.handleJob(r.queue.items[1]!);
    expect((await r.storage.threads.get(ran.threadId))!.state).toBe('COMPLETED');
  });
});
